[agent] Implement Backhaul STA Steering

VladyslavTupikin opened the pull request:
## Description

According to UML diagram need to do an implementation for the _**Backhaul STA Steering**_:

1. Add implementation for the **_associate()_** method in __bwl/dwpal__: it should send `NETWORK_ENABLE `message to the **_wpa suplicant_**.
2. Add implementation for the **_Backhaul STA Steering Request_** message handler in the **_backhaul_manager_**: method should parse TLV and if channel from TLV correct, trigger the **_associate()_** method.
3. Add implementation for the **_Backhaul STA Steering Response_** message creator method in the **_backhaul_manager_**.
4. Add reaction in Backhaul manager for getting `ENABLED `event from the **_wpa supplicant_** in the **_backhaul_manager_**: **_backhaul_manager_** should check specific flag which was turned on in the handler, and **if it is true**, send  **_Backhaul STA Steering Response_** message to the controller.
5. Add implementation for the **_Backhaul STA Steering Response_** message in the controller code.

https://jira.prplfoundation.org/browse/PPM-168

Commit:
agent: backhaul_manager: Create handler for Backhaul STA Steering Request

According to the UML diagram need to implement
handler for Backhaul STA Steering Request in the
backhaul manager wih name handle_backhaul_steering_request.

Add implementation for handler, add new method for handler 1905 messages,
add BACKHAUL_STEERING_REQUEST_MESSAGE type to bus_subscribe method.

handle_backhaul_steering_request should parse tlvBackhaulSteeringRequest,
trigger the associate method from current HAL and send ACK message
to the controller.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
agent: backhaul_manager: Add method for sending Backhaul STA Steering Response

Need to add method which creates Backhaul STA Steering Response
message with filled up parameters target_bssid, target_bssid,
result_code in tlvBackhaulSteeringResponse.

Add implementation for method create_backhaul_steering_response.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
agent: backhaul_manager: Add logic for sending Backhaul STA Steering Response

Need to add sending Backhaul STA Steering Response
message when backhaul_thread recieves ENABLED event
from the wpa_supplicant.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
controller: son_master: Add handler for BACKHAUL_STA_STEERING_RESPONSE_MESSAGE

Need to add handler for the BACKHAUL_STA_STEERING_RESPONSE_MESSAGE.

Add handler with printing the type of message, MID, result
of BACKHAUL_STA_STEERING_RESPONSE_MESSAGE and the error code.
Also add sendind ACK message to the controller according to the
paragraph 6 section 12 Easy Mesh R1 specification.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

arnout commented on common/beerocks/bwl/dwpal/sta_wlan_hal_dwpal.cpp:
```diff
@@ -242,7 +242,42 @@ bool sta_wlan_hal_dwpal::connect(const std::string &ssid, const std::string &pas
+    auto bssid_str = tlvf::mac_to_string(bssid);
+
+    LOG(DEBUG) << " Associate iface " << get_iface_name() << " to BSSID = " << bssid_str
+               << ", Channel = " << int(channel);
```
As mentioned in #1424 this function isn't needed. It was wrong anyway because it didn't do any steering.

VladyslavTupikin replied:
This commit removed.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -4458,6 +4464,49 @@ bool backhaul_manager::handle_slave_channel_selection_response(ieee1905_1::CmduM
+                                                        const std::string &src_mac)
+{
+    const auto mid = cmdu_rx.getMessageId();
+    LOG(DEBUG) << "Received BACKHAUL_STA_STEERING message, mid=" << std::hex << int(mid);
```
```suggestion
    LOG(DEBUG) << "Received BACKHAUL_STA_STEERING message, mid=" << std::hex << mid;
```
Explicit conversion to int is only needed for enums and for int8_t. It used to be needed for uint8_t as well, but that has been solved by adding an explicit operator<< overload.

And anyway, mid is uint16_t so this was never needed.

VladyslavTupikin replied:
fixed

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -4458,6 +4464,49 @@ bool backhaul_manager::handle_slave_channel_selection_response(ieee1905_1::CmduM
+    auto bh_sta_steering_req = cmdu_rx.getClass<wfa_map::tlvBackhaulSteeringRequest>();
+    if (!bh_sta_steering_req) {
+        LOG(ERROR) << "Failed cmdu_rx.getClass<wfa_map::tlvBackhaulSteeringRequest>(), mid="
+                   << std::hex << int(mid);
```
```suggestion
                   << std::hex << mid;
```

VladyslavTupikin replied:
fixed

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -4458,6 +4464,49 @@ bool backhaul_manager::handle_slave_channel_selection_response(ieee1905_1::CmduM
+
+    auto bh_sta_steering_req = cmdu_rx.getClass<wfa_map::tlvBackhaulSteeringRequest>();
+    if (!bh_sta_steering_req) {
+        LOG(ERROR) << "Failed cmdu_rx.getClass<wfa_map::tlvBackhaulSteeringRequest>(), mid="
```
```suggestion
        LOG(WARNING) << "Failed cmdu_rx.getClass<wfa_map::tlvBackhaulSteeringRequest>(), mid="
```
Since this message comes from outside, it is not an "error" (in the sense that something is wrong with our system). Rather , it is invalid input.

Doesn't matter much of course, but we should probably do this right.

VladyslavTupikin replied:
accepted!

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -4458,6 +4464,49 @@ bool backhaul_manager::handle_slave_channel_selection_response(ieee1905_1::CmduM
+        return false;
+    }
+
+    auto associate = active_hal->associate(bssid, channel);
```
```suggestion
    auto associate = active_hal->roam(bssid, channel);
```

The proper way is probably to use the state machine instead. That said, maybe it's not needed, and it's way more efficient if we can roam directly here rather than going through the state machine.

VladyslavTupikin replied:
Roger that.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -4458,6 +4464,49 @@ bool backhaul_manager::handle_slave_channel_selection_response(ieee1905_1::CmduM
+        LOG(ERROR) << "cmdu creation of type ACK_MESSAGE, has failed";
+        return false;
+    }
+    LOG(DEBUG) << "sending ACK message to the originator, mid=" << std::hex << int(mid);
```
```suggestion
    LOG(DEBUG) << "sending ACK message to the originator, mid=" << std::hex << mid;
```

VladyslavTupikin replied:
fixed

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -4507,6 +4506,31 @@ bool backhaul_manager::handle_backhaul_steering_request(ieee1905_1::CmduMessageR
+    const wfa_map::tlvBackhaulSteeringResponse::eResultCode &result_code)
+{
+    auto cmdu_tx_header = m_expected_backhaul_sta_steering.response.create(
+        0, ieee1905_1::eMessageType::BACKHAUL_STEERING_RESPONSE_MESSAGE);
```
Since we are not building up the message from different responses here but we're just expecting a single one, there's no need to keep a response buffer in m_expected_backhaul_sta_steering. We can simply use cmdu_tx here.
```suggestion
    auto cmdu_tx_header = cmdu_tx.create(
        0, ieee1905_1::eMessageType::BACKHAUL_STEERING_RESPONSE_MESSAGE);
```

VladyslavTupikin replied:
accepted

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -886,6 +886,14 @@ bool backhaul_manager::backhaul_fsm_main(bool &skip_select)
+                m_expected_backhaul_sta_steering.backhaul_sta_steering_enable = false;
+
+                send_cmdu_to_bus(m_expected_backhaul_sta_steering.response, controller_bridge_mac,
+                                 bridge_info.mac);
```
Here, we should also create the response.

The bssid is active_hal->get_bssid(). Funnily enough, the station MAC address is not so easy to get: you have to get it from active_hal->get_iface_name() using get_iface_mac().

VladyslavTupikin replied:
solved

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1794,6 +1799,25 @@ bool master_thread::handle_cmdu_1905_topology_response(const std::string &src_ma
+{
+    auto mid = cmdu_rx.getMessageId();
+    LOG(DEBUG) << "Received BACKHAUL_STA_STEERING_MESSAGE from " << src_mac << ", mid=" << std::hex
+               << int(mid);
```
```suggestion
               << mid;
```

VladyslavTupikin replied:
fixed

arnout reviewed (approved):
Although some comments are pretty major, since we don't have any way to test this yet there's no point gating the merge on that. So accepting.

vitalii-komisarenko commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -4580,6 +4579,41 @@ bool backhaul_manager::handle_backhaul_steering_request(ieee1905_1::CmduMessageR
+    const wfa_map::tlvBackhaulSteeringResponse::eResultCode &result_code)
+{
+    auto cmdu_tx_header =
+        cmdu_tx.create(0, ieee1905_1::eMessageType::BACKHAUL_STEERING_RESPONSE_MESSAGE);
```
Is it OK to use message ID 0?
Often message ID of the request and the response should be the same.
Certification test 4.9.1 does not check it. Can it change in the future?

arnout replied:
Yes it has to be a new MID. The request is answered with an ACK that has the same MID as the request, while the response uses a new MID. It's a bit confusing...

vitalii-komisarenko reviewed (approved):
Some commit messages are too long.
"target_bssid" is menioned two times in the commit message of 7454c656c19467af759282066cc7f793f247f056.
Other than that looks OK.

Commit:
agent: backhaul_manager: Add channel validation in handler

Need to add channel validation for Backhaul STA Steering
Request handler. If channel is invalid - send ACK message
and Backhaul STA Steering Response message with FAILURE
code.

Remove TODO commentary and add changes described above.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -4505,6 +4511,58 @@ bool backhaul_manager::handle_slave_channel_selection_response(ieee1905_1::CmduM
+              Response with the error code without trigerring HAL.
+    */
+
+    auto active_hal = get_wireless_hal();
+    if (!active_hal) {
+        LOG(ERROR) << "Couldn't get active HAL";
+        return false;
+    }
```
This can actually happen if we have wired backhaul. I'm not sure what to do in that case, but it would be good to at least add a TODO here. I think we should still send an ACK anyway. Sending ACK should be done before getting active hal.

VladyslavTupikin replied:
Added TODO above getting HAL. Sending ACK message moved upper than getting active HAL.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -4505,6 +4511,58 @@ bool backhaul_manager::handle_slave_channel_selection_response(ieee1905_1::CmduM
+
+    LOG(DEBUG) << "Sending ACK message to the originator, mid=" << std::hex << mid;
+    return send_cmdu_to_bus(cmdu_tx, controller_bridge_mac, bridge_info.mac);
+}
```
Sending ACK should already be done before doing the roam, because we have to do it even if the roam fails.

VladyslavTupikin replied:
Accepted

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -4563,6 +4562,41 @@ bool backhaul_manager::handle_backhaul_steering_request(ieee1905_1::CmduMessageR
 }
 
+bool backhaul_manager::create_backhaul_steering_response(
+    const wfa_map::tlvBackhaulSteeringResponse::eResultCode &result_code)
```
I think we should have as parameter the error code rather than the result code (see below).

VladyslavTupikin replied:
accepted!

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -4563,6 +4562,41 @@ bool backhaul_manager::handle_backhaul_steering_request(ieee1905_1::CmduMessageR
+    auto interface = active_hal->get_iface_name();
+    get_iface_mac(interface, sta_mac);
+
+    LOG(DEBUG) << "Interface: " << interface << "MAC: " << tlvf::mac_to_string(sta_mac);
```
```suggestion
    LOG(DEBUG) << "Interface: " << interface << "MAC: " << sta_mac;
```

VladyslavTupikin replied:
fixed

vitalii-komisarenko reviewed (approved):
My review is identical to previous one: https://github.com/prplfoundation/prplMesh/pull/1445#pullrequestreview-438281483
Mostly related to commit messages.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -4563,6 +4562,41 @@ bool backhaul_manager::handle_backhaul_steering_request(ieee1905_1::CmduMessageR
+
+    bh_steering_resp_tlv->target_bssid()         = tlvf::mac_from_string(active_hal->get_bssid());
+    bh_steering_resp_tlv->backhaul_station_mac() = sta_mac;
+    bh_steering_resp_tlv->result_code()          = result_code;
```
If result code is failure, we also have to provide an error code TLV with the error code. See the 4th paragraph of section 12.

So I think it's better to pass the error code directly, and to set the result code (and include the error TLV) depending on whether the error code is 0 or not.

VladyslavTupikin replied:
You are right. Added creation for ErrorCodeTlv

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -597,6 +601,8 @@ class backhaul_manager : public btl::transport_socket_thread {
 
     sExpectedChannelSelection m_expected_channel_selection;
 
+    bool backhaul_sta_steering_enable = false;
```
This change is unrelated and not needed I think.

VladyslavTupikin replied:
Moved to commit with usage.

arnout reviewed (commented):
Only reviewed the first 2 commits, but I have to take a break now :-)

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -969,6 +969,16 @@ bool backhaul_manager::backhaul_fsm_main(bool &skip_select)
         m_eth_link_up       = network_utils::linux_iface_is_up_and_running(m_sConfig.wire_iface);
         FSM_MOVE_STATE(OPERATIONAL);
+
+        if (backhaul_sta_steering_enable) {
```
This needs a comment IMO. Something like:

```
// This event may come as a result of enabling the backhaul, but also as a result
// of steering. *Only* in case it was the result of steering, we need to send a steering
// response.
```

VladyslavTupikin replied:
Accepted

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -4551,13 +4561,21 @@ bool backhaul_manager::handle_backhaul_steering_request(ieee1905_1::CmduMessageR
-            TODO: Need to add sending Backhaul STA Steering Response with the FAILURE error code.
-        */
+        auto response = create_backhaul_steering_response(
+            wfa_map::tlvBackhaulSteeringResponse::eResultCode::FAILURE);
```
As mentioned earlier, in case of failure we also need to give an error code. We have the choice of BACKHAUL_STEERING_REQUEST_REJECTED_CANNOT_OPERATE_ON_CHANNEL_SPECIFIED, BACKHAUL_STEERING_REQUEST_REJECTED_TARGET_BSS_SIGNAL_NOT_SUITABLE or BACKHAUL_STEERING_REQUEST_AUTHENTICATION_OR_ASSOCIATION_REJECTED. Actually none of these are applicable here because if the roam() call failed, it just means there's a problem with the connection to wpa_supplicant. I think BACKHAUL_STEERING_REQUEST_REJECTED_TARGET_BSS_SIGNAL_NOT_SUITABLE is the closest to be accurate.

VladyslavTupikin replied:
Accepted.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1794,6 +1799,25 @@ bool master_thread::handle_cmdu_1905_topology_response(const std::string &src_ma
+
+    LOG(DEBUG) << "BACKHAUL_STA_STEERING_MESSAGE result_code: "
+               << int(tlv_backhaul_sta_steering_resp->result_code());
+
```
The controller should respond with an ACK.

VladyslavTupikin replied:
Added sending ACK message to the agent and fixed commit message.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -4554,6 +4544,33 @@ bool backhaul_manager::handle_backhaul_steering_request(ieee1905_1::CmduMessageR
+        send_cmdu_to_bus(cmdu_tx, controller_bridge_mac, bridge_info.mac);
+
+        auto response = create_backhaul_steering_response(
+            wfa_map::tlvBackhaulSteeringResponse::eResultCode::FAILURE);
```
Here the error code must be BACKHAUL_STEERING_REQUEST_REJECTED_CANNOT_OPERATE_ON_CHANNEL_SPECIFIED.

VladyslavTupikin replied:
Fixed

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -4554,6 +4544,33 @@ bool backhaul_manager::handle_backhaul_steering_request(ieee1905_1::CmduMessageR
+        LOG(WARNING) << "Invalid channel number";
+
+        LOG(DEBUG) << "Sending ACK message to the originator, mid=" << std::hex << mid;
+        send_cmdu_to_bus(cmdu_tx, controller_bridge_mac, bridge_info.mac);
```
This is independent on whether the channel is valid or not, so it can be done before is_valid_channel.

VladyslavTupikin replied:
Accepted

arnout reviewed (approved):
Except for including an error code TLV in the response in case of failure, none of my comments are really critical. Also there's no need for me to re-review.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -4517,6 +4535,139 @@ bool backhaul_manager::handle_slave_channel_selection_response(ieee1905_1::CmduM
+        return false;
+    }
+
+    error_code_tlv->reason_code() = error_code;
```
This should only be added if error_code != 0.

VladyslavTupikin replied:
fixed
